server/game/game.js

```diff
@@ -829,14 +824,76 @@ class Game extends EventEmitter {
         card.controller = newController;
     }
 
+    watch(socketId, user) {
+        if(!this.allowSpectators) {
+            return false;
+        }
+
+        this.players[user.username] = new Spectator(socketId, user);
+        this.addMessage('{0} has joined the game as a spectator', user.username);
+
+        return true;
+    }
+
+    join(socketId, user) {
+        if(this.started || _.values(this.getPlayers()).length === 2) {
+            return false;
+        }
+
+        this.players[user.username] = new Player(socketId, user, this.owner === user.username, this);
+
+        return true;
+    }
+
+    isEmpty() {
+        return _.all(this.players, player => player.disconnected || player.left);
+    }
+
+    leave(playerName) {
+        var player = this.players[playerName];
+
+        if(!player) {
+            return;
+        }
+
+        this.addMessage('{0} has left the game', player);
```

Comment:
One idea i had recently, that your implementation now makes easier, is to notify specifically if it's a spectator that's left/disconnected - dunno about anyone else but i find myself having to check if it was my opponent who left or not :)